### ISSUE TITLE
Can't resolve modules with Next js 13.1.5

### ISSUE THREAD

**Issue by t0rbik** [WalletConnect/walletconnect-monorepo#1908]
**Describe the bug**
When trying to use WalletConnectConnector with wagmi in latest Next.js 13.1.5 with experimental appDir enabled compiler throws several warnings, saying: Module not found: Can't resolve <package>

**Version**
Looking at package-lock there're:
"@walletconnect/ethereum-provider": "^1.8.0",
"@walletconnect/universal-provider": "^2.2.1",

**package.json**
{
  "name": "my-app",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "ethers": "^5.7.2",
    "next": "^13.1.5",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "wagmi": "^0.11.0"
  },
  "devDependencies": {
    "@types/node": "18.0.1",
    "@types/react": "18.0.14",
    "@types/react-dom": "18.0.5",
    "eslint": "8.19.0",
    "eslint-config-next": "12.2.0",
    "typescript": "^4.9.4"
  }
}

**To Reproduce**
[minimal reproducible example](https://replit.com/@IgorTorbin/walletconnect-issue?v=1)
1. Go to minimal reproducible example
2. Click on edit in workspace or open workspace. You need to see console.
3. Run the app
4. Watch console for errors
5. See the error:
```
warn  - ./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/db.js
Module not found: Can't resolve 'lokijs' in '/home/runner/walletconnect-issue/node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js'

Import trace for requested module:
./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/db.js
./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/index.js
./node_modules/@walletconnect/keyvaluestorage/dist/cjs/index.js
./node_modules/@walletconnect/sign-client/node_modules/@walletconnect/core/dist/index.cjs.js
./node_modules/@walletconnect/sign-client/dist/index.cjs.js
./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
./node_modules/@wagmi/connectors/dist/walletConnect.js
./node_modules/@wagmi/core/dist/connectors/walletConnect.js
./node_modules/wagmi/dist/connectors/walletConnect.js
./app/(components)/WagmiProvider.tsx

./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
Module not found: Can't resolve 'encoding' in '/home/runner/walletconnect-issue/node_modules/cross-fetch/node_modules/node-fetch/lib'

Import trace for requested module:
./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@walletconnect/jsonrpc-http-connection/dist/cjs/http.js
./node_modules/@walletconnect/jsonrpc-http-connection/dist/cjs/index.js
./node_modules/@walletconnect/ethereum-provider/dist/cjs/index.js
./node_modules/@wagmi/connectors/dist/walletConnect.js
./node_modules/@wagmi/core/dist/connectors/walletConnect.js
./node_modules/wagmi/dist/connectors/walletConnect.js
./app/(components)/WagmiProvider.tsx

./node_modules/pino/lib/tools.js
Module not found: Can't resolve 'pino-pretty' in '/home/runner/walletconnect-issue/node_modules/pino/lib'

Import trace for requested module:
./node_modules/pino/lib/tools.js
./node_modules/pino/pino.js
./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
./node_modules/@wagmi/connectors/dist/walletConnect.js
./node_modules/@wagmi/core/dist/connectors/walletConnect.js
./node_modules/wagmi/dist/connectors/walletConnect.js
./app/(components)/WagmiProvider.tsx
```

**Expected behavior**
No warns in the console

**Comment by amzfcy** [comment]
Will this warning persist in the local environment?

**Comment by codetaskforce** [comment]
I had a similar problem if not the same, after attempting a next 12 > 13 transition today. 

No error showed on the client console just only on VS console. UI appears to work as should so is this a @ wagmi/@ walletconnect update needed to combat the below error? 

```
warn  - ./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/db.js
Module not found: Can't resolve 'lokijs' in '...\y-next13\node_modules\@walletconnect\keyvaluestorage\dist\cjs\node-js'

Import trace for requested module:
./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/db.js
./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/index.js
./node_modules/@walletconnect/keyvaluestorage/dist/cjs/index.js
./node_modules/@walletconnect/core/dist/index.cjs.js
./node_modules/@walletconnect/sign-client/dist/index.cjs.js
./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
./node_modules/@wagmi/connectors/dist/walletConnect.js
./node_modules/@wagmi/core/dist/connectors/walletConnect.js
./node_modules/wagmi/dist/connectors/walletConnect.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./src/components/header.tsx
./src/app/layout.tsx

./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
Module not found: Can't resolve 'encoding' in '...\y-next13\node_modules\cross-fetch\node_modules\node-fetch\lib'

Import trace for requested module:
./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@walletconnect/jsonrpc-http-connection/dist/cjs/http.js
./node_modules/@walletconnect/jsonrpc-http-connection/dist/cjs/index.js
./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
./node_modules/@wagmi/connectors/dist/walletConnect.js
./node_modules/@wagmi/core/dist/connectors/walletConnect.js
./node_modules/wagmi/dist/connectors/walletConnect.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./src/components/header.tsx
./src/app/layout.tsx

./node_modules/pino/lib/tools.js
Module not found: Can't resolve 'pino-pretty' in '...\y-next13\node_modules\pino\lib'

Import trace for requested module:
./node_modules/pino/lib/tools.js
./node_modules/pino/pino.js
./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
./node_modules/@wagmi/connectors/dist/walletConnect.js
./node_modules/@wagmi/core/dist/connectors/walletConnect.js
./node_modules/wagmi/dist/connectors/walletConnect.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./src/components/header.tsx
./src/app/layout.tsx
```

Anyway this cured it for me:-
```
yarn add pino-pretty lokijs@1.x encoding
```

[👨‍💻](https://emojipedia.org/man-technologist/)

**Comment by t0rbik** [comment]
> Will this warning persist in the local environment?

I encountered it initially in local environment. Didn't try adding those packages as @codetaskforce solution states, probably can add those to peer deps?

**Comment by Dope-Ass-Ninja** [comment]
I'm having this same issue in my local environment. Doesn't seem to affect production builds but the error occurs in the build log as well.

**Comment by solvman** [comment]
the same issue... @codetaskforce solution seem to fix the dependency issue.

**Comment by Xiawpohr** [comment]
I find another solution if you don't want to install dependencies not related to your project.
```
// in next.config.js
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    appDir: true,
  },
  webpack: (config) => {
    config.externals.push("pino-pretty", "lokijs", "encoding");
    return config;
  },
};

module.exports = nextConfig;
```

It works for me. No error shows in the console.

**Comment by notamicrodose** [comment]
Same issue here... the issue might be related to the bundler not being able to resolve the dependencies properly in a browser environment. However @codetaskforce 's solution works for me as well, so no point in investigating further lol

**Comment by boogoogle** [comment]
- I think it is the problem of Next.js, it can't divide client dependencies and server dependencies in deep. 
- I have tried the same code in Vite + React+TS project, it works with no error. 
- finally thanks  to @Xiawpohr , finally we fix the shit of Next.js by ourselves

**Comment by glitch-txs** [comment]
It's seems like you guys found a workaround, if the issue persists and you'd like a different solution we can check on wagmi's WalletConnect connector.

tldr: Next.js will render your app first on server side, this is triggered when the init function is called in a node.js environment. If  you're using ethereum-provider you can call it inside useEffect (which won't run on server side) or by checking that the window object is defined before initializing the ethereum provider.

Closing for now as completed.

**Comment by o-az** [comment]
@Xiawpohr's solution didn't work for me. This is what worked:

```ts
const nextConfig = {
  webpack: (config, context) => {
    if (config.plugins) {
      config.plugins.push(
        new context.webpack.IgnorePlugin({
          resourceRegExp: /^(lokijs|pino-pretty|encoding)$/,
        }),
      )
    }
    return config
  },
}
```

[webpack `IgnorePlugin` API docs](https://webpack.js.org/plugins/ignore-plugin/)

**Comment by transmental** [comment]
> I had a similar problem if not the same, after attempting a next 12 > 13 transition today.
> 
> No error showed on the client console just only on VS console. UI appears to work as should so is this a @ wagmi/@ walletconnect update needed to combat the below error?
> 
> ```
> warn  - ./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/db.js
> Module not found: Can't resolve 'lokijs' in '...\y-next13\node_modules\@walletconnect\keyvaluestorage\dist\cjs\node-js'
> 
> Import trace for requested module:
> ./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/db.js
> ./node_modules/@walletconnect/keyvaluestorage/dist/cjs/node-js/index.js
> ./node_modules/@walletconnect/keyvaluestorage/dist/cjs/index.js
> ./node_modules/@walletconnect/core/dist/index.cjs.js
> ./node_modules/@walletconnect/sign-client/dist/index.cjs.js
> ./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
> ./node_modules/@wagmi/connectors/dist/walletConnect.js
> ./node_modules/@wagmi/core/dist/connectors/walletConnect.js
> ./node_modules/wagmi/dist/connectors/walletConnect.js
> ./node_modules/@rainbow-me/rainbowkit/dist/index.js
> ./src/components/header.tsx
> ./src/app/layout.tsx
> 
> ./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
> Module not found: Can't resolve 'encoding' in '...\y-next13\node_modules\cross-fetch\node_modules\node-fetch\lib'
> 
> Import trace for requested module:
> ./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
> ./node_modules/cross-fetch/dist/node-ponyfill.js
> ./node_modules/@walletconnect/jsonrpc-http-connection/dist/cjs/http.js
> ./node_modules/@walletconnect/jsonrpc-http-connection/dist/cjs/index.js
> ./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
> ./node_modules/@wagmi/connectors/dist/walletConnect.js
> ./node_modules/@wagmi/core/dist/connectors/walletConnect.js
> ./node_modules/wagmi/dist/connectors/walletConnect.js
> ./node_modules/@rainbow-me/rainbowkit/dist/index.js
> ./src/components/header.tsx
> ./src/app/layout.tsx
> 
> ./node_modules/pino/lib/tools.js
> Module not found: Can't resolve 'pino-pretty' in '...\y-next13\node_modules\pino\lib'
> 
> Import trace for requested module:
> ./node_modules/pino/lib/tools.js
> ./node_modules/pino/pino.js
> ./node_modules/@walletconnect/universal-provider/dist/index.cjs.js
> ./node_modules/@wagmi/connectors/dist/walletConnect.js
> ./node_modules/@wagmi/core/dist/connectors/walletConnect.js
> ./node_modules/wagmi/dist/connectors/walletConnect.js
> ./node_modules/@rainbow-me/rainbowkit/dist/index.js
> ./src/components/header.tsx
> ./src/app/layout.tsx
> ```
> 
> Anyway this cured it for me:-
> 
> ```
> yarn add pino-pretty lokijs@1.x encoding
> ```
> 
> [👨‍💻](https://emojipedia.org/man-technologist/)

this fixed it for me

**Comment by matthewlilley** [comment]
> It's seems like you guys found a workaround, if the issue persists and you'd like a different solution we can check on wagmi's WalletConnect connector.
> 
> tldr: Next.js will render your app first on server side, this is triggered when the init function is called in a node.js environment. If you're using ethereum-provider you can call it inside useEffect (which won't run on server side) or by checking that the window object is defined before initializing the ethereum provider.
> 
> Closing for now as completed.

Yeah, it would be nice to fix this for the wagmi connector. It's only going to run client-side so these deps and ignoring them shouldn't be necessary.
